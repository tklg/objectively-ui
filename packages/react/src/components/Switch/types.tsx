import { ClassAttributes, InputHTMLAttributes, Ref } from 'react'
import { CommonColor, CommonSize } from 'src/types/props'

type InputPropsType = ClassAttributes<HTMLInputElement> & InputHTMLAttributes<HTMLInputElement>

export interface SwitchProps extends Omit<InputPropsType, 'value' | 'defaultValue' | 'size' | 'onChange'> {
  label?: string;
  inputRef?: Ref<HTMLInputElement>;
  size?: CommonSize;
  color?: CommonColor;
  inputProps?: object;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>, checked: boolean) => void;
}
