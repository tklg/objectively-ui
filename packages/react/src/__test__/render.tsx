import { CssVarsProvider } from 'src/providers'
import { render, RenderOptions } from '@testing-library/react'
import { FC, PropsWithChildren, ReactElement } from 'react'

const AllTheProviders: FC<PropsWithChildren> = ({children}) => {
  return <CssVarsProvider>{children}</CssVarsProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
