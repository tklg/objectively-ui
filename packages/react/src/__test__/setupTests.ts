import '@testing-library/jest-dom/extend-expect'
import matchMediaMock from 'src/__test__/matchMedia.mock'
import { createSerializer } from '@emotion/jest'

matchMediaMock()

// configures @emotion/jest to not insert styles
expect.addSnapshotSerializer(createSerializer({
  includeStyles: false,
}))
