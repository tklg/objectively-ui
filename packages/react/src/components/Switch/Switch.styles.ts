import { css } from '@emotion/react'
import { focusOutlineStylesWithoutElement } from 'src/components/Accessibility.styles'
import { ColorTheme } from 'src/types/ColorTheme'
import { CommonSize } from 'src/types/props'

const sizeFactor = 0.6
const widthFactor = 1.6

export const switchStyles = (theme: ColorTheme, size: CommonSize) => css({
  height: theme.size[size],
  display: 'flex',
  padding: `calc((${theme.size[size]} - (${theme.size[size]} * ${sizeFactor})) / 2) 0`,
})

export const switchTrackStyles = (theme: ColorTheme, size: CommonSize) => css({
  position: 'relative',
  display: 'inline-block',
  height: `calc(${theme.size[size]} * ${sizeFactor})`,
  width: `calc(${theme.size[size]} * ${sizeFactor * widthFactor})`,
  backgroundColor: theme.colors.hover,
  borderRadius: theme.radii[size],
  cursor: 'pointer',
  outline: 'none',
  transition: [
    `background ${theme.transitions.duration.fast} ${theme.transitions.function.default}`,
  ].join(','),

  '&:focus-visible:before': {
    content: '\'\'',
    position: 'absolute',
    height: '100%',
    width: '100%',
    top: 0,
    left: 0,
    ...focusOutlineStylesWithoutElement(theme),
  },

  'input:checked + &': {
    backgroundColor: theme.colors.accentPrimary.value,
  },

  'input:checked:disabled + &': {
    backgroundColor: theme.colors.accentPrimary.shadow,
  },
})

export const switchHandleStyles = (theme: ColorTheme, size: CommonSize) => css({
  position: 'absolute',
  backgroundColor: theme.colors.accentPrimary.contrastText,
  height: `calc(${theme.size[size]} * ${sizeFactor} - ${theme.spacing.xxs} * 2)`,
  width: `calc(${theme.size[size]} * ${sizeFactor} - ${theme.spacing.xxs} * 2)`,
  borderRadius: `calc(${theme.radii[size]} - ${theme.spacing.xxs})`,
  top: theme.spacing.xxs,
  left: theme.spacing.xxs,
  transform: 'translateX(0)',
  transition: [
    `transform ${theme.transitions.duration.fast} ${theme.transitions.function.default}`,
  ].join(','),

  'input:checked + label > &': {
    transform: `translateX(calc(${theme.size[size]} * ${sizeFactor * widthFactor} - ${theme.size[size]} * ${sizeFactor}))`,
  },

  'input:disabled + label > &': {
    backgroundColor: theme.colors.textDisabled,
  },
})

export const switchLabelStyles = (theme: ColorTheme, size: CommonSize) => css({
  display: 'inline-block',
  marginLeft: theme.spacing.xs,
  lineHeight: `calc(${theme.size[size]} * ${sizeFactor})`,
  fontSize: theme.typography.size[size],
})
