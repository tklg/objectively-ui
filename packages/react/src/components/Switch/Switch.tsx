import { ChangeEvent, forwardRef, useCallback, useEffect, useId, useRef, useState } from 'react'
import { switchHandleStyles, switchLabelStyles, switchStyles, switchTrackStyles } from 'src/components/Switch/Switch.styles'
import { SwitchProps } from 'src/components/Switch/types'
import { useTheme } from 'src/hooks'
import { useKeyboardEvents } from 'src/hooks/useKeyboardEvents'
import { buildClassName } from 'src/utils/buildClassName'
import { KEYS } from 'src/utils/keys'
import { mergeRefs } from 'src/utils/mergeRefs'

const ELEMENT_NAME = 'Switch'

export const Switch = forwardRef<HTMLInputElement, SwitchProps>(({
  inputRef: externalInputRef,
  label,
  checked,
  onChange,
  className: _className,
  id: _id,
  readOnly,
  disabled,
  size = 'md',
  color = 'primary',
  inputProps = {},
  ...props
}, ref) => {
  const fallbackId = useId()
  const theme = useTheme()
  const inputRef = useRef<HTMLInputElement>()
  const id = _id || fallbackId
  const [internalChecked, setInternalChecked] = useState(checked)
  const internalCheckedRef = useRef(checked)
  const className = buildClassName(ELEMENT_NAME, {
    readOnly,
    disabled,
    size,
    color,
  }, _className)
  const trackClassName = buildClassName(`${ELEMENT_NAME}Track`)
  const handleClassName = buildClassName(`${ELEMENT_NAME}Handle`)
  const labelClassName = buildClassName(`${ELEMENT_NAME}Label`)

  useEffect(() => {
    if (checked !== internalCheckedRef.current) {
      setInternalChecked(checked)
    }
  }, [checked])

  const handleChange = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    const newState = !internalChecked
    internalCheckedRef.current = newState
    setInternalChecked(newState)
    onChange?.(e, newState)
  }, [internalChecked, onChange])

  const handleKeydown = useCallback(() => {
    inputRef.current?.click()
  }, [])

  const keyboardProps = useKeyboardEvents([KEYS.Space, KEYS.Enter], handleKeydown, 'keydown')

  return (
    <div
      {...props}
      className={className}
      css={switchStyles(theme, size)}
      ref={ref}
    >
      <input
        ref={mergeRefs(inputRef, externalInputRef)}
        {...inputProps}
        checked={internalChecked}
        onChange={handleChange}
        readOnly={readOnly}
        disabled={disabled}
        type='checkbox'
        hidden
        tabIndex={-1}
        aria-hidden
        id={id}
      />
      <label
        className={trackClassName}
        css={switchTrackStyles(theme, size)}
        role='switch'
        aria-checked={internalChecked ?? 'false'}
        tabIndex={disabled ? -1 : 0}
        htmlFor={id}
        {...keyboardProps}
      >
        <div className={handleClassName} css={switchHandleStyles(theme, size)} />
      </label>
      {label && (
        <label htmlFor={id} className={labelClassName} css={switchLabelStyles(theme, size)}>{label}</label>
      )}
    </div>
  )
})

if (process.env.NODE_ENV !== 'production') {
  Switch.displayName = ELEMENT_NAME
}
