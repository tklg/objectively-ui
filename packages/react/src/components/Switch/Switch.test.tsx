import { Switch } from 'src/components/Switch/Switch'
import { fireEvent, render, screen } from 'src/__test__/render'

describe('Switch', () => {
  it('renders a switch', async () => {
    render(
      <Switch data-testid='switch' inputProps={{ 'data-testid': 'checkbox' }} />,
    )

    expect(screen.getByRole('switch')).toHaveAttribute('aria-checked', 'false')
    expect(screen.getByTestId('checkbox')).not.toBeChecked()

    fireEvent.click(screen.getByRole('switch'))

    expect(screen.getByRole('switch')).toHaveAttribute('aria-checked', 'true')
    expect(screen.getByTestId('checkbox')).toBeChecked()

    expect(screen.getByTestId('switch')).toMatchSnapshot()
  })

  it('renders a disabled switch', async () => {
    render(
      <Switch inputProps={{ 'data-testid': 'checkbox' }} disabled />,
    )

    expect(screen.getByRole('switch')).toHaveAttribute('tabindex', '-1')
    expect(screen.getByTestId('checkbox')).toBeDisabled()
  })

  it('fires onChange', async () => {
    const handleChange = jest.fn()

    render(
      <Switch onChange={handleChange} />,
    )

    fireEvent.click(screen.getByRole('switch'))
    expect(handleChange).toHaveBeenCalledWith(expect.anything(), true)

    fireEvent.click(screen.getByRole('switch'))
    expect(handleChange).toHaveBeenCalledWith(expect.anything(), false)
  })

  it('respects external state', async () => {
    const handleChange = jest.fn()

    const { rerender } = render(
      <Switch checked={false} onChange={handleChange} inputProps={{ 'data-testid': 'checkbox' }} />,
    )

    expect(handleChange).not.toHaveBeenCalled()
    expect(screen.getByRole('switch')).toHaveAttribute('aria-checked', 'false')
    expect(screen.getByTestId('checkbox')).not.toBeChecked()

    rerender(
      <Switch checked={true} onChange={handleChange} inputProps={{ 'data-testid': 'checkbox' }} />,
    )

    expect(handleChange).not.toHaveBeenCalled()
    expect(screen.getByRole('switch')).toHaveAttribute('aria-checked', 'true')
    expect(screen.getByTestId('checkbox')).toBeChecked()
  })
})
