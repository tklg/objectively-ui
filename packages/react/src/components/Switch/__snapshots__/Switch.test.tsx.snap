// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders a switch 1`] = `
<div
  class="ObjUI-Switch ObjUI-Switch-sizeMd ObjUI-Switch-colorPrimary emotion-0"
  data-testid="switch"
>
  <input
    aria-hidden="true"
    data-testid="checkbox"
    hidden=""
    id=":r0:"
    tabindex="-1"
    type="checkbox"
  />
  <label
    aria-checked="true"
    class="ObjUI-SwitchTrack emotion-1"
    for=":r0:"
    role="switch"
    tabindex="0"
  >
    <div
      class="ObjUI-SwitchHandle emotion-2"
    />
  </label>
</div>
`;
